src/vsr/replica.zig

```diff
@@ -7252,6 +7253,18 @@ pub fn ReplicaType(
                 });
                 return;
             }
+            // To avoid falsely contributing to the durability of the current checkpoint, replicas
+            // syncing table blocks should not send prepare_oks for prepares past (and including)
+            // the next checkpoint's trigger.
+            if (!self.sync_content_done() and self.op_checkpoint() > 0 and
+                header.op >= vsr.Checkpoint.prepare_max_for_checkpoint(self.op_checkpoint()).?)
```

Comment:
Hm, this feels overly strict to me. There are only a pipeline worth of preapres for which `header.op <= vsr.Checkpoint.prepare_max_for_checkpoint(self.op_checkpoint()).?` is true. 

So, I think practically this condition is equivalent to "don't ever ok prepares if you are syncing content": there are about 1000 prepares in a checkpoint, and we are allowed to ack only 8 of them. 

I _think_ this could be relaxed quite a bit. Assume that we are syncing content, and we are at checkpoint `C`. Now, anything between `C` and `C + trigger + pipeline` is safe to ack. 

`C + trigger` is trivially safe, as those things are even committed. 

`trigger <-> pipeline`is ok because that's before `prepare_max(C)` and doesn't signal that we have reached `C` (in other words, _even_ if we were on `C - checkpoint_interval`, we could still acked those prepares).

Now, something like `C + trigger + pipeline + 1`  is unsafe to ack -- the primary can look at that, and decide that we have checkpoint `C`, and decide that `C` is durable on the cluster, and then evict some log message, compormisign durability.

But I think anything beyond `C + trigger + pipeline + pipeline` is safe to ack again? Because if we see something in that range, we know that `C + trigger + pipeline + 1` is already committed, and that means that the checkpoint `C` _is_ actually durable?

Reply:
> anything beyond C + trigger + pipeline + pipeline is safe to ack again

Beautiful! Indeed this can be relaxed by a *LOT*, since if we see an op beyond C + trigger + pipeline + pipeline, it is guaranteed that (C + trigger + pipeline + 1) is committed. Updating this condition right away.

Reply:
Would be great to double-check the logic with replica test ^^

Reply:
Writing a replica test to reproduce this issue unveiled that this logic that we discussed is correct but doesn't address the issue at hand! I have pushed the replica test as part of https://github.com/tigerbeetle/tigerbeetle/pull/2221/commits/f7a82d22e3adab532ea4472890ed1fbcd3195a7f for you to peruse, although it is a little janky right now as one test that fails on main and passes on this branch right now (branched by a hardcoded flag). 

While it is correct to withhold prepares between _[C<sub>n</sub> + trigger + pipeline, C<sub>n</sub> + trigger + 2 * pipeline]_ if we're not sure about durability of C<sub>n</sub>, the main triggering condition in this issue is that the syncing replica allows the primary to move on to the *next* checkpoint. Specifically, with the help of the syncing replica R0, R2 advances it's checkpoint to C<sub>n+1</sub>. Consequently, R2 is not able to help R1 with a corruption in its checkpoint (since it has a newer version of that block). R0 isn't able to help R1 with the corruption since it hasn't finished syncing its tables.

I think it makes sense to revive the logic that we started off with, wherein a syncing replica withholds prepare_oks for prepares past the trigger of the _next_ checkpoint. I've implemented both of these changes together in https://github.com/tigerbeetle/tigerbeetle/pull/2221/commits/6b589c58607f27aac4e167fd59bb7a665b3d3b2a, what do you think?

Reply:
In terms of this pr: let's land the frist part of witholding prepares, as the first scenario above is a convincing demonstartion why we need it.

For the second part, the _last_ scenario I feel shows that we'll need to do more work here

Reply:
Also, the last scenario indeed seems like something that we would have to discuss more elaborately to come up with a general solution from the problem.

However @matklad, in the meantime, what are your thoughts about pushing the current solution (syncing replicas don't contribute prepare_oks for checkpoint advancement) as a temporary mitigation? It has been VOPR green for 12223 runs till now, and it does address the failing seed that this PR aims to solve.

Reply:
>One question though, in the first scenario, wouldn't withholding prepare_oks between [Cn + trigger + pipeline, Cn + trigger + 2 * pipeline] keep R1 & R2 at op_checkpoint=39

Ah, yeah, you are right!

>what are your thoughts about pushing the current solution (syncing replicas don't contribute prepare_oks for checkpoint advancement) as a temporary mitigation?

I'd rather not to! As I learned from my failed attempts to fix the old state-sync protocol, incremental "this fixes this particular case" adjustments are not necessary a part of the final solution, and might actually create extra complexity when you'd have to back them out.

Reply:
> might actually create extra complexity when you'd have to back them out.

Ack, that does make sense. Let's solve it once and for all, without any mitigations on the way. :)

I will do away with the "syncing replicas don't contribute prepare_oks for checkpoint advancement" logic then! Also will change the description for this PR and to reflect what it is _actually_ trying to solve, i.e. scenario 1 from https://github.com/tigerbeetle/tigerbeetle/pull/2221#discussion_r1723213288 (with small corrections).